st2common/st2common/router.py

```diff
@@ -0,0 +1,386 @@
+            LOG.debug('Route registered: %+6s %s', route.conditions['method'][0], route.routepath)
+
+    def match(self, req):
+        path = req.path
+
+        if len(path) > 1 and path.endswith('/'):
+            path = path[:-1]
+
+        match = self.routes.match(path, req.environ)
+
+        if match is None:
+            raise NotFoundException('No route matches "%s" path' % req.path)
+
+        # To account for situation when match may return multiple values
+        try:
+            path_vars = match[0]
+        except KeyError:
+            path_vars = match
+
+        path = path_vars.pop('_api_path')
+        method = path_vars.pop('_api_method')
+        endpoint = self.spec['paths'][path][method]
+
+        return endpoint, path_vars
+
+    def __call__(self, req):
+        """
+        The method is invoked on every request and shows the lifecycle of the request received from
+        the middleware.
+
+        Although some middleware may use parts of the API spec, it is safe to assume that if you're
+        looking for the particular spec property handler, it's most  likely a part of this method.
+
+        At the time of writing, the only property being utilized by middleware was `x-log-result`.
+        """
+        endpoint, path_vars = self.match(req)
+
+        context = copy.copy(getattr(self, 'mock_context', {}))
+
+        # Handle security
+        if 'security' in endpoint:
+            security = endpoint.get('security')
+        else:
+            security = self.spec.get('security', [])
+
+        if self.auth and security:
+            try:
+                auth_resp = None
+                security_definitions = self.spec.get('securityDefinitions', {})
+                for statement in security:
+                    declaration, options = statement.copy().popitem()
+                    definition = security_definitions[declaration]
+
+                    if definition['type'] == 'apiKey':
+                        if definition['in'] == 'header':
+                            token = req.headers.get(definition['name'])
+                        elif definition['in'] == 'query':
+                            token = req.GET.get(definition['name'])
+                        else:
+                            token = None
+
+                        if token:
+                            if auth_resp:
+                                raise auth_exc.MultipleAuthSourcesError(
+                                    'Only one of Token or API key expected.')
+
+                            auth_func = op_resolver(definition['x-operationId'])
+                            auth_resp = auth_func(token)
+
+                            context['user'] = User.get_by_name(auth_resp.user)
+
+                if 'user' not in context:
+                    raise auth_exc.NoAuthSourceProvidedError('One of Token or API key required.')
+            except (auth_exc.NoAuthSourceProvidedError,
```

Comment:
I though about using middleware for that but decided against it at the end. I see authentication an inalienable part of the application and if we decide to extend it, we have other means for that.